File: tests/DesktopGrass.Win2D.Tests/ConfigTests.cs
using DesktopGrass.Win2D;
using Xunit;

namespace DesktopGrass.Win2D.Tests;

public sealed class ConfigTests
{
    private static string FreshConfigPath(string name)
    {
        string dir = Path.Combine(Directory.GetCurrentDirectory(), ".copilot-scratch", "win2d-config-tests", name);
        if (Directory.Exists(dir)) Directory.Delete(dir, recursive: true);
        Directory.CreateDirectory(dir);
        return Path.Combine(dir, "config.json");
    }

    [Fact]
    public void MissingFile_YieldsDefaults_AndWritesTemplate()
    {
        string path = FreshConfigPath("missing");
        Assert.False(File.Exists(path));

        AppConfig cfg = Config.Load(path);

        Assert.Equal(Config.TargetFpsDefault, cfg.TargetFps);
        Assert.Equal(Config.BladeDensityDefault, cfg.BladeDensity, 6);

        // A default file should now exist and re-read identically (it is JSONC).
        Assert.True(File.Exists(path));
        AppConfig reread = Config.Load(path);
        Assert.Equal(Config.TargetFpsDefault, reread.TargetFps);
        Assert.Equal(Config.BladeDensityDefault, reread.BladeDensity, 6);
    }

    [Fact]
    public void ValidValues_AreParsed()
    {
        string path = FreshConfigPath("valid");
        File.WriteAllText(path, "{ \"version\": 1, \"targetFps\": 60, \"bladeDensity\": 1.5 }");

        AppConfig cfg = Config.Load(path);
        Assert.Equal(60, cfg.TargetFps);
        Assert.Equal(1.5, cfg.BladeDensity, 6);
    }

    [Fact]
    public void OutOfRangeValues_AreClamped()
    {
        string path = FreshConfigPath("clamp");

        File.WriteAllText(path, "{ \"targetFps\": 1000, \"bladeDensity\": 99.0 }");
        AppConfig high = Config.Load(path);
        Assert.Equal(Config.TargetFpsMax, high.TargetFps);
        Assert.Equal(Config.BladeDensityMax, high.BladeDensity, 6);

        File.WriteAllText(path, "{ \"targetFps\": 0, \"bladeDensity\": 0.0 }");
        AppConfig low = Config.Load(path);
        Assert.Equal(Config.TargetFpsMin, low.TargetFps);
        Assert.Equal(Config.BladeDensityMin, low.BladeDensity, 6);
    }

    [Fact]
    public void JsoncCommentsAndTrailingCommas_AreTolerated()
    {
        string path = FreshConfigPath("jsonc");
        File.WriteAllText(path,
            "{\n" +
            "  // a comment\n" +
            "  \"targetFps\": 24, /* inline */\n" +
            "  \"bladeDensity\": 2.0,\n" +
            "}\n");

        AppConfig cfg = Config.Load(path);
        Assert.Equal(24, cfg.TargetFps);
        Assert.Equal(2.0, cfg.BladeDensity, 6);
    }

    [Fact]
    public void MalformedFile_FallsBackToDefaults_AndIsPreserved()
    {
        string path = FreshConfigPath("malformed");
        File.WriteAllText(path, "{ not valid json ");

        AppConfig cfg = Config.Load(path);
        Assert.Equal(Config.TargetFpsDefault, cfg.TargetFps);
        Assert.Equal(Config.BladeDensityDefault, cfg.BladeDensity, 6);

        // The user's broken file must be left untouched for them to fix.
        Assert.Equal("{ not valid json ", File.ReadAllText(path));
    }

    [Fact]
    public void MissingKeys_FallBackToPerKeyDefaults()
    {
        string path = FreshConfigPath("partial");
        File.WriteAllText(path, "{ \"targetFps\": 45 }");

        AppConfig cfg = Config.Load(path);
        Assert.Equal(45, cfg.TargetFps);
        Assert.Equal(Config.BladeDensityDefault, cfg.BladeDensity, 6);
        Assert.Equal(Config.SwaySpeedDefault, cfg.SwaySpeed, 6);
        Assert.Equal(Config.SwayAmplitudeDefault, cfg.SwayAmplitude, 6);
    }

    [Fact]
    public void Keys_AreMatchedCaseInsensitively()
    {
        string path = FreshConfigPath("case-insensitive");
        File.WriteAllText(path,
            "{ \"TargetFps\": 60, \"BLADEDENSITY\": 1.5, " +
            "\"SwaySpeed\": 0.5, \"swayamplitude\": 2.0 }");

        AppConfig cfg = Config.Load(path);
        Assert.Equal(60, cfg.TargetFps);
        Assert.Equal(1.5, cfg.BladeDensity, 6);
        Assert.Equal(0.5, cfg.SwaySpeed, 6);
        Assert.Equal(2.0, cfg.SwayAmplitude, 6);
    }

    [Fact]
    public void SwayKnobs_Parse_Clamp_AndRejectNonFinite()
    {
        string path = FreshConfigPath("sway");

        // Defaults when absent.
        File.WriteAllText(path, "{ }");
        AppConfig cfg = Config.Load(path);
        Assert.Equal(Config.SwaySpeedDefault, cfg.SwaySpeed, 6);
        Assert.Equal(Config.SwayAmplitudeDefault, cfg.SwayAmplitude, 6);

        // Valid values parsed.
        File.WriteAllText(path, "{ \"swaySpeed\": 0.5, \"swayAmplitude\": 2.0 }");
        cfg = Config.Load(path);
        Assert.Equal(0.5, cfg.SwaySpeed, 6);
        Assert.Equal(2.0, cfg.SwayAmplitude, 6);

        // Out-of-range clamped to bounds.
        File.WriteAllText(path, "{ \"swaySpeed\": 99.0, \"swayAmplitude\": -5.0 }");
        cfg = Config.Load(path);
        Assert.Equal(Config.SwaySpeedMax, cfg.SwaySpeed, 6);
        Assert.Equal(Config.SwayAmplitudeMin, cfg.SwayAmplitude, 6);

        // Non-finite (inf from overflow) falls back to default, never poisons the sim.
        File.WriteAllText(path, "{ \"swaySpeed\": 1e999, \"swayAmplitude\": 1e999 }");
        cfg = Config.Load(path);
        Assert.Equal(Config.SwaySpeedDefault, cfg.SwaySpeed, 6);
        Assert.Equal(Config.SwayAmplitudeDefault, cfg.SwayAmplitude, 6);
    }
}

File: tests/DesktopGrass.Native.Tests/src/config_tests.cpp
#include "TestHelpers.h"
#include "Config.h"

#include <filesystem>
#include <fstream>
#include <sstream>
#include <string>

using namespace desktopgrass;

namespace {

std::filesystem::path test_config_path(const char* name) {
    std::filesystem::path dir = std::filesystem::current_path()
        / ".copilot-scratch"
        / "native-config-tests"
        / name;
    std::error_code ec;
    std::filesystem::remove_all(dir, ec);
    std::filesystem::create_directories(dir);
    return dir / "config.json";
}

void write_text(const std::filesystem::path& path, const std::string& text) {
    std::filesystem::create_directories(path.parent_path());
    std::ofstream file(path, std::ios::binary | std::ios::trunc);
    file << text;
}

std::string read_text(const std::filesystem::path& path) {
    std::ifstream file(path, std::ios::binary);
    std::ostringstream buffer;
    buffer << file.rdbuf();
    return buffer.str();
}

} // namespace

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace DesktopGrassNativeTests
{
TEST_CLASS(ConfigTests)
{
public:
TEST_METHOD(ConfigMissingFileYieldsDefaultsAndWritesATemplate) {
    const std::filesystem::path path = test_config_path("missing");
    Assert::IsFalse(std::filesystem::exists(path));

    const config::Config cfg = config::LoadConfig(path.wstring());

    Assert::IsTrue(cfg.targetFps == config::kTargetFpsDefault);
    Assert::IsTrue(cfg.bladeDensity == Near(config::kBladeDensityDefault));

    // A default file should have been created and be re-readable (it is JSONC).
    Assert::IsTrue(std::filesystem::exists(path));
    const config::Config reread = config::LoadConfig(path.wstring());
    Assert::IsTrue(reread.targetFps == config::kTargetFpsDefault);
    Assert::IsTrue(reread.bladeDensity == Near(config::kBladeDensityDefault));
}

TEST_METHOD(ConfigValidValuesAreParsed) {
    const std::filesystem::path path = test_config_path("valid");
    write_text(path, "{ \"version\": 1, \"targetFps\": 60, \"bladeDensity\": 1.5 }");

    const config::Config cfg = config::LoadConfig(path.wstring());
    Assert::IsTrue(cfg.targetFps == 60);
    Assert::IsTrue(cfg.bladeDensity == Near(1.5));
}

TEST_METHOD(ConfigOutOfRangeValuesAreClamped) {
    const std::filesystem::path path = test_config_path("clamp");
    write_text(path, "{ \"targetFps\": 1000, \"bladeDensity\": 99.0 }");

    config::Config cfg = config::LoadConfig(path.wstring());
    Assert::IsTrue(cfg.targetFps == config::kTargetFpsMax);
    Assert::IsTrue(cfg.bladeDensity == Near(config::kBladeDensityMax));

    write_text(path, "{ \"targetFps\": 0, \"bladeDensity\": 0.0 }");
    cfg = config::LoadConfig(path.wstring());
    Assert::IsTrue(cfg.targetFps == config::kTargetFpsMin);
    Assert::IsTrue(cfg.bladeDensity == Near(config::kBladeDensityMin));
}

TEST_METHOD(ConfigOversizedIntegerValuesFallBackSafely) {
    const std::filesystem::path path = test_config_path("oversized-integer");

    write_text(path, "{ \"targetFps\": 1e300 }");
    config::Config cfg = config::LoadConfig(path.wstring());
    Assert::IsTrue(cfg.targetFps == config::kTargetFpsDefault);

    write_text(path, "{ \"targetFps\": -1e300 }");
    cfg = config::LoadConfig(path.wstring());
    Assert::IsTrue(cfg.targetFps == config::kTargetFpsDefault);
}

TEST_METHOD(ConfigJSONCCommentsAndTrailingCommasAreTolerated) {
    const std::filesystem::path path = test_config_path("jsonc");
    write_text(path,
        "{\n"
        "  // a comment\n"
        "  \"targetFps\": 24, /* inline */\n"
        "  \"bladeDensity\": 2.0,\n"  // trailing comma below
        "}\n");

    const config::Config cfg = config::LoadConfig(path.wstring());
    Assert::IsTrue(cfg.targetFps == 24);
    Assert::IsTrue(cfg.bladeDensity == Near(2.0));
}

TEST_METHOD(ConfigMalformedFileFallsBackToDefaultsAndIsPreserved) {
    const std::filesystem::path path = test_config_path("malformed");
    write_text(path, "{ not valid json ");

    const config::Config cfg = config::LoadConfig(path.wstring());
    Assert::IsTrue(cfg.targetFps == config::kTargetFpsDefault);
    Assert::IsTrue(cfg.bladeDensity == Near(config::kBladeDensityDefault));

    // The user's (broken) file must be left untouched for them to fix.
    Assert::IsTrue(read_text(path) == "{ not valid json ");
}

TEST_METHOD(ConfigMissingKeysFallBackToPerKeyDefaults) {
    const std::filesystem::path path = test_config_path("partial");
    write_text(path, "{ \"targetFps\": 45 }");

    const config::Config cfg = config::LoadConfig(path.wstring());
    Assert::IsTrue(cfg.targetFps == 45);
    Assert::IsTrue(cfg.bladeDensity == Near(config::kBladeDensityDefault));
    Assert::IsTrue(cfg.swaySpeed == Near(config::kSwaySpeedDefault));
    Assert::IsTrue(cfg.swayAmplitude == Near(config::kSwayAmplitudeDefault));
}

TEST_METHOD(ConfigKeysAreMatchedCaseInsensitively) {
    const std::filesystem::path path = test_config_path("case-insensitive");
    write_text(path,
        "{ \"TargetFps\": 60, \"BLADEDENSITY\": 1.5, "
        "\"SwaySpeed\": 0.5, \"swayamplitude\": 2.0 }");

    const config::Config cfg = config::LoadConfig(path.wstring());
    Assert::IsTrue(cfg.targetFps == 60);
    Assert::IsTrue(cfg.bladeDensity == Near(1.5));
    Assert::IsTrue(cfg.swaySpeed == Near(0.5));
    Assert::IsTrue(cfg.swayAmplitude == Near(2.0));
}

TEST_METHOD(ConfigSwayKnobsParseClampAndRejectNonFinite) {
    const std::filesystem::path path = test_config_path("sway");

    // Defaults when absent.
    write_text(path, "{ }");
    config::Config cfg = config::LoadConfig(path.wstring());
    Assert::IsTrue(cfg.swaySpeed == Near(config::kSwaySpeedDefault));
    Assert::IsTrue(cfg.swayAmplitude == Near(config::kSwayAmplitudeDefault));

    // Valid values parsed.
    write_text(path, "{ \"swaySpeed\": 0.5, \"swayAmplitude\": 2.0 }");
    cfg = config::LoadConfig(path.wstring());
    Assert::IsTrue(cfg.swaySpeed == Near(0.5));
    Assert::IsTrue(cfg.swayAmplitude == Near(2.0));

    // Out-of-range clamped to bounds.
    write_text(path, "{ \"swaySpeed\": 99.0, \"swayAmplitude\": -5.0 }");
    cfg = config::LoadConfig(path.wstring());
    Assert::IsTrue(cfg.swaySpeed == Near(config::kSwaySpeedMax));
    Assert::IsTrue(cfg.swayAmplitude == Near(config::kSwayAmplitudeMin));

    // Non-finite (inf from overflow) falls back to default, never poisons the sim.
    write_text(path, "{ \"swaySpeed\": 1e999, \"swayAmplitude\": 1e999 }");
    cfg = config::LoadConfig(path.wstring());
    Assert::IsTrue(cfg.swaySpeed == Near(config::kSwaySpeedDefault));
    Assert::IsTrue(cfg.swayAmplitude == Near(config::kSwayAmplitudeDefault));
}
};
}
